src/kernel-install/ukify.py

```diff
@@ -0,0 +1,187 @@
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+def parse_args():
+    p = argparse.ArgumentParser(
+        description='Build and sign Unified Kernel Images')
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
+
+    p.add_argument('--os-release',
+                   type=pathlib.Path)
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
+                   type=pathlib.Path)
+
+    p.add_argument('--efi-arch',
+                   choices=('ia32', 'x64', 'arm', 'aa64', 'riscv64'))
+
+    p.add_argument('--stub',
+                   type=pathlib.Path)
+
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
+
+    p.add_argument('--measure',
+                   action='store_true')
+
+    args = p.parse_args()
+
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
+
+    if args.os_release is None:
+        p = pathlib.Path('/etc/os-release')
+        if not p.exists():
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
+
+    if args.output is None:
+        suffix = '.efi' if args.key else '.unsigned.efi'
+        args.output = args.linux.name + suffix
+
+    return args
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name == 'output':
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        open(value).close()
+
+
+def make_uki(args):
+    files = []  # a holder for temporary files
+    sections = []
+
+    def add_section(name, content, offset):
+        if not content:
+            return
+
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            files.append(tmp)
+            content = tmp.name
+
+        sections.append((name, content, offset))
+
+    add_section('osrel',   args.os_release,    0x20_000)
+    add_section('cmdline', args.cmdline,       0x30_000)
+    add_section('dtb',     args.devicetree,    0x40_000)
+    add_section('splash',  args.splash,       0x100_000)
+    add_section('linux',   args.linux,      0x2_000_000)
+    add_section('initrd',  args.initrd,     0x3_000_000)
```

Comment:
It's a terrible practice that these values have been hard coded since forever. These really need to be picked based on actual size of the payload added and the largest VMA of the stub image. Especially if native PE image creation ever comes into the mix as that might generate large VMA offsets, meaning these fixed values make the final image too big to fit in memory.

Also, as long as there is fallback support for EFI handover protocol the linux section should really get some special attention:
 - it ought to be marked as executable with --set-section-flags
 - the decompressor expects enough free space after the kernel proper, so there needs to be breathing room after it from the PE header (SizeOfCode and/or SizeOfUninitializedData). The previous hard coded values were just lucky (or just chosen carefully). Really, I think the linux section should just go last. (Apparently, uncompressed aarch64 kernels can end up being too large for the default gap.)

Reply:
Here is why: https://github.com/systemd/systemd/issues/24202

And it wasn't just us hitting issues recently in CI, but people have been hitting this on older stubs as well (though, why people use that part of github to discuss things instead of filing a bug report is beyond me. it would've saved me so much time solving the above bug): https://github.com/systemd/systemd/commit/0fa2cac4f0cdefaf1addd7f1fe0fd8113db9360b#commitcomment-55421176

Reply:
The Arch Wiki was also updated with some bash to calculate the offsets.

https://wiki.archlinux.org/title/Unified_kernel_image#Manually

Reply:
I tried to use`--set-section-flags`, but `objcopy` doesn't like that:

```console
$ objcopy /usr/lib/systemd/boot/efi/linuxx64.efi.stub --add-section .osrel=/etc/os-release --change-section-vma .osrel=0x4c38000 --add-section .cmdline=/tmp/tmpcmdline0p7h5fqp --change-section-vma .cmdline=0x4c39000 --add-section .pcrpkey=tpm2-pcr-public.pem --change-section-vma .pcrpkey=0x4c3a000 --add-section .initrd=/boot/08a5690a2eed47cf92ac0a5d2e3cf6b0/6.0.5-300.fc37.x86_64/initrd --change-section-vma .initrd=0x4c3b000 --add-section .uname=/tmp/tmpunameah3_h5y3 --change-section-vma .uname=0x8c1f000 --add-section .test=/tmp/tmptestp4jozmd4 --change-section-vma .test=0x8c20000 --add-section .test2=/tmp/tmptest252whqfze --change-section-vma .test2=0x8c21000 --add-section .pcrsig=/tmp/tmppcrsigllow37su --change-section-vma .pcrsig=0x8c22000 --add-section .linux=/lib/modules/6.0.5-300.fc37.x86_64/vmlinuz --change-section-vma .linux=0x8c23000 --set-section-flags .linux=contents,alloc,load,readonly,code /tmp/uki4f4pp6ao
objcopy: /tmp/uki4f4pp6ao: can't add section '.linux': file in wrong format
```

Reply:
What objcopy is this? It works for me on arch with 2.39.0. Afaik you only need to pass `--set-section-flags .linux=code,readonly`.

Ultimately, this isn't too important as is only really needed for the deprecated handover path and only out of paranoia. And stub assembly so far never set this as section as executable and nobody ever complained.

Reply:
I guess I tried to pass too many flags. With `--set-section-flags .linux=code,readonly` it indeed works.